packages/react-admin/src/sideEffect/saga/i18n.js

```diff
@@ -0,0 +1,33 @@
+import { all, call, put, select, takeLatest } from 'redux-saga/effects';
+import {
+    CHANGE_LOCALE,
+    USER_LOGOUT,
+    changeLocale,
+    changeLocaleSuccess,
+    changeLocaleFailure,
+} from '../../actions';
+import { getLocale } from '../../reducer';
+
+export default i18nProvider => {
+    function* loadMessages(action) {
+        const locale = action.payload;
+        try {
+            const messages = yield call(i18nProvider, locale);
+            yield put(changeLocaleSuccess(locale, messages));
+        } catch (err) {
+            yield put(changeLocaleFailure(locale));
+        }
+    }
+    function* loadInitialMessages() {
+        const initialLocale = yield select(getLocale);
+        yield put(changeLocale(initialLocale));
+    }
+
+    return function*() {
+        yield all([
+            takeLatest(CHANGE_LOCALE, loadMessages),
+            takeLatest(USER_LOGOUT, loadInitialMessages),
+            call(loadInitialMessages),
```

Comment:
I'm afraid that this one will cause an initial double render of the `<Admin>` children. Can you check that it doesn't?

Reply:
Yes I had the same concern, my observations are that it does when the messages are loaded async, but synchronous I don't see this effect. In my custom app I have a loader which `races` the `CHANGE_LOCALE_SUCCESS` and `CHANGE_LOCALE_FAILURE` before displaying `react-admin`, maybe this should be added?

Reply:
Let's summarize the problems: 

1. The loading of the messages can be async to avoid loading all messages at startup
2. We don't want to block the rendering of the app while it waits for an async message fetch
3. But the app requires a default locale, loaded synchronously, for the first render
4. The locale can depend on the user (a custom `authSaga` can dispatch a `CHANGE_LOCALE` upon login).

The solution, IMO, is to implement two actions in the `i18nProvider`:

- `GET_DEFAULT_MESSAGES`: must be synchronous, is called by the `<Admin>` component, then passed to the i18n *reducer* to initialize the default i18n state
- `CHANGE_LOCALE`: can be asynchronous, and is called by the i18n *sagas*

That way, no need to reset messages upon logout, because the state has a default set of i18n messages.

The way I see it, the i18nProvider could still be a simple function taking the locale as only argument, and returning a promise for messages (if the locale argument is defined), or directly messages (if no locale is passed). Or, it could be a more structured function, similar to the `dataProvider`, taking type and params, as follows:

```js
const i18nProvider = (messages, defaultMessages) => (type, params) => {
    if (type === 'GET_DEFAULT_MESSAGES') {
        return defaultMessages;
    }
    if (type === 'CHANGE_LOCALE') {
        return messages[params.locale];
    }
    throw new Error('Undefined action type', type); 
}
```

Having to pass both a `messages` and a `defaultMessages` props to the `<Admin>` component will be cumbersome, so I guess we can take advantage of the BC break in 2.0 and ask only for one parameter: the `i18nProvider`. The default `i18nProvider` would simply be:

```js
import englishMessages from 'ra-language-english'
const i18nProvider = (type, params) => englishMessages;
```

What do you think?